L80-csharp-load-balancing.md

```diff
@@ -0,0 +1,259 @@
+        void IDictionary<string, object>.Add(string key, object? value);
+        bool IDictionary<string, object>.ContainsKey(string key);
+        object? IDictionary<string, object>.get_Item(string key);
+        bool IDictionary<string, object>.Remove(string key);
+        void IDictionary<string, object>.set_Item(string key, object? value);
+        bool IDictionary<string, object>.TryGetValue(string key, out object? value);
+        IEnumerator<KeyValuePair<string, object?>> IEnumerable<KeyValuePair<string, object>>.GetEnumerator();
+        IEnumerator IEnumerable.GetEnumerator();
+    }
+    public readonly struct BalancerAttributesKey<TValue> {
+        public BalancerAttributesKey(string key);
+        public string Key { get; }
+    }
+    public class BalancerState {
+        public BalancerState(ConnectivityState connectivityState, SubChannelPicker picker);
+        public ConnectivityState ConnectivityState { get; }
+        public SubChannelPicker Picker { get; }
+    }
+    public class ChannelState {
+        public ChannelState(AddressResolverResult resolverState, BalancerAttributes options);
+        public BalancerAttributes Options { get; set; }
+        public AddressResolverResult ResolverState { get; set; }
+    }
+    public class CompleteContext {
+        public CompleteContext();
+        public Exception? Error { get; set; }
+        public Metadata? Trailers { get; set; }
+    }
+    public class DnsAddressResolver : AddressResolver {
+        public DnsAddressResolver(Uri address, ILoggerFactory loggerFactory);
+        public override Task RefreshAsync(CancellationToken cancellationToken);
+        public override IDisposable Subscribe(IObserver<AddressResolverResult> observer);
+        protected override void Dispose(bool disposing);
+    }
+    public class DnsAddressResolverFactory : AddressResolverFactory {
+        public DnsAddressResolverFactory(ILoggerFactory loggerFactory);
+        public override string Name { get; }
+        public override AddressResolver Create(Uri address, AddressResolverOptions options);
+    }
+    public abstract class LoadBalancer : IDisposable {
+        protected LoadBalancer();
+        public void Dispose();
+        public abstract void ResolverError(Exception exception);
+        public abstract void UpdateChannelState(ChannelState state);
+        public abstract void UpdateSubChannelState(SubChannel subChannel, SubChannelState state);
+        protected virtual void Dispose(bool disposing);
+    }
+    public abstract class LoadBalancerFactory {
+        protected LoadBalancerFactory();
+        public abstract string Name { get; }
+        public abstract LoadBalancer Create(IConnectionController controller, IDictionary<string, object> options);
+    }
+    public class PickContext {
+        public PickContext();
+        public HttpRequestMessage? Request { get; set; }
+        public bool WaitForReady { get; set; }
```

Comment:
The picker should not be aware of this.  It should be handled by the channel.

Basically, the channel should do the following based on the pick result:
- If the picker returned PICK_COMPLETE, send the call to the subchannel that the picker returned.
- If the picker returned PICK_QUEUE, queue the pick to try again later.
- If the picker returned PICK_DROP, fail the RPC with the status indicated by the picker.
- If the picker returned PICK_FAILED:
  - If the call is wait_for_ready=false, then fail the RPC with the status returned by the picker.
  - Otherwise, queue the pick to try again later.

All queued picks will be retried the next time the LB policy returns a new picker to the channel.

Reply:
Removed.